tests/test_interfaces.py

```diff
@@ -1,79 +1,102 @@
+        operator.samples, operator.shape, n_coils=operator.n_coils, smaps=operator.smaps
+    )
+
+
+@fixture(scope="module")
+def image_data(operator):
+    """Generate a random image."""
+    if operator.smaps is None:
+        img = np.random.rand(operator.n_coils, *operator.shape).astype(
+            operator.cpx_dtype
+        )
+    elif operator.smaps is not None and operator.n_coils > 1:
+        img = np.random.rand(*operator.shape).astype(operator.cpx_dtype)
 
-    def case_radial2D(self, Nc=10, Ns=10):
-        pass
+    img += 1j * np.random.randn(*img.shape).astype(operator.cpx_dtype)
+    return img
 
-    def case_radial3D(self, Nc=10, Ns=10, Nr=10, expansion="rotation"):
-        pass
 
-    def case_grid1D(self, N=256):
-        freq_1d = sp.fft.fftfreq(N)
-        return freq_1d.reshape(-1, 1), (N,)
+@fixture(scope="module")
+def kspace_data(operator):
+    """Generate a random kspace data."""
+    kspace = np.random.randn(operator.n_coils, operator.n_samples).astype(
+        operator.cpx_dtype
+    )
+    return kspace
 
-    def case_grid2D(self, N=16):
-        """Create a cartesian grid of frequencies locations."""
-        freq_1d = sp.fft.fftfreq(N)
-        freq_2d = np.stack(np.meshgrid(freq_1d, freq_1d), axis=-1)
-        return freq_2d.reshape(-1, 2), (N, N)
 
-    def case_grid3D(self, N=16):
-        """Create a cartesian grid of frequencies locations."""
-        freq_1d = sp.fft.fftfreq(N)
-        freq_3d = np.stack(np.meshgrid(freq_1d, freq_1d, freq_1d), axis=-1)
-        return freq_3d.reshape(-1, 3), (N, N, N)
+def test_interfaces_accuracy_forward(operator, image_data, nfft_ref_op):
+    """Compare the interface to the raw NUDFT implementation."""
+    kspace_nufft = operator.op(image_data).squeeze()
+    kspace_ref = nfft_ref_op.op(image_data).squeeze()
+    # FIXME: check with complex values ail
+    assert np.percentile(abs(kspace_nufft - kspace_ref), 95) < 1e-4
+    assert np.max(abs(kspace_nufft - kspace_ref)) < 1e-3
 
 
-@parametrize_with_cases(
-    "kspace_grid, shape",
-    cases=[
-        CasesTrajectories.case_grid1D,
-        CasesTrajectories.case_grid2D,
-    ],  # 3D is ignored (too much possibility for the reordering)
-)
-def test_ndft_matrix(kspace_grid, shape):
-    """Test that  the ndft matrix is a good matrix for doing fft."""
-    ndft_op = RawNDFT(kspace_grid, shape, explicit_matrix=True)
-    # Create a random image
-    fft_matrix = [None] * len(shape)
-    for i in range(len(shape)):
-        fft_matrix[i] = sp.fft.fft(np.eye(shape[i]), norm="ortho")
-    fft_mat = fft_matrix[0]
-    if len(shape) == 2:
-        fft_mat = fft_matrix[0].flatten()[:, None] @ fft_matrix[1].flatten()[None, :]
-        fft_mat = (
-            fft_mat.reshape(shape * 2)
-            .transpose(2, 0, 1, 3)
-            .reshape((np.prod(shape),) * 2)
-        )
-    assert np.allclose(ndft_op._fourier_matrix, fft_mat)
+def test_interfaces_accuracy_backward(operator, kspace_data, nfft_ref_op):
```

Comment:
I dont understand how are these functions called for test? Am I missing some API for parameterize?

Reply:
see https://smarie.github.io/python-pytest-cases/ for the details.
TLDR: the tuple (kspace_grid, shape) is generate by each case function selected via `cases` argument. This is a simpler interface to pytest's fixtures

Reply:
the operator argument is auto-magically detected as a fixture by pytest (because there is function with the same name and a fixture decorator). All the parametrization is done on the operator, and then each test is run with it.

Reply:
Ahhh cool, yes got it!